### PR TITLE
Study equality

### PR DESCRIPTION
**Pull request by Kris2ffer**
Added equals method for all internal models.
Input models are meant to be converted before testing equality.

### CODE REVIEW

File: RestApi/Sepes.RestApi/Model/Pod.cs [hunk 1]
```diff
@@ -36,10 +45,32 @@ public class Pod
...
+        public override int GetHashCode()
+        {
+            var hash = new HashCode();
+            hash.Add(id);
+            hash.Add(name);
+            hash.Add(studyId);
+            hash.Add(allowAll);
+            hash.Add(incoming);
+            hash.Add(outgoing);
+            hash.Add(users);
```
Comment (gjersvik) [inline RestApi/Sepes.RestApi/Model/Pod.cs]: You can't simply take the hash of lists. As that link to standard list equality. Best to not use them as part of the hash.

File: RestApi/Sepes.RestApi/Model/Pod.cs [hunk 2]
```diff
@@ -36,10 +45,32 @@ public class Pod
...
+            hash.Add(locked);
+            hash.Add(loaded);
```
Comment (gjersvik) [inline RestApi/Sepes.RestApi/Model/Pod.cs]: Do not put thing in the hash that is not in Equals. The other way around is fine. The hash rule is: If hash != hash then object must be not equal object. If hash == hash then object may be equal object.

File: RestApi/Sepes.RestApi/Model/Study.cs [hunk 3]
```diff
@@ -44,5 +47,25 @@ public class Study
...
+                   Enumerable.SequenceEqual(suppliers, study.suppliers) &&
+                   Enumerable.SequenceEqual(datasets, study.datasets) &&
+                   archived == study.archived &&
+                   Enumerable.SequenceEqual(userIds, study.userIds) &&
+                   Enumerable.SequenceEqual(datasetIds, study.datasetIds);
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(studyName, studyId, pods, sponsors, suppliers, archived, userIds, datasetIds);
```
Comment (gjersvik) [inline RestApi/Sepes.RestApi/Model/Study.cs]: Cant use pods, sponsors, suppliers, userIds or datasetIds as part of hash. See previous issue.

File: RestApi/Sepes.RestApi.Tests/Model/StudyTest.cs [hunk 4]
```diff
@@ -36,5 +37,51 @@ public void ConstructorInput()
...
+            var studySmall = new Study("My study", 123);
+            var studySmall2 = new Study("My study", 123, new HashSet<Pod>());
+            var podsSmall = new HashSet<Pod>();
+            podsSmall.Add(new Pod(1, "test", 1));
+            var studySmall3 = new Study("My study", 123, podsSmall);
+
+            Assert.True(studySmall.Equals(studySmall2));
+            Assert.Equal(1, studySmall3.pods.Count);
+            Assert.False(studySmall.Equals(studySmall3));
+        }
```
Comment (gjersvik) [inline RestApi/Sepes.RestApi.Tests/Model/StudyTest.cs]: Please split up into multiple tests. This one is to big by an order of magnitude for an Equals test. There should be 3 lines in each. Create a, Create b and compare.

**Review by gjersvik — CHANGES_REQUESTED** [review]
Some issues with hashes.

**Review by Unicus-Tom — APPROVED** [review]

**Review by gjersvik — APPROVED** [review]
Looks a lot better.

**Pull request merged**